modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/GridDhtTxFinishFuture.java

```diff
@@ -509,7 +509,7 @@ private boolean finish(boolean commit,
                         + n.id()));
                 }
                 else {
-                    cctx.io().send(n, req, tx.ioPolicy());
+                    cctx.io().send(n, cctx.snapshotMgr().wrapMessage(req, tx.cutId()), tx.ioPolicy());
```

Comment:
Can we implement some kind of plug in message wrapper inside of `GridCacheIoManager` ?
```
    private volatile Function<GridCacheMessage, GridCacheMessage> msgProc;

    public void send(ClusterNode node, GridCacheMessage msg, byte plc) throws IgniteCheckedException {
        assert !node.isLocal() : node;
        Function<GridCacheMessage, GridCacheMessage> msgProc0 = msgProc
        if (msgProc0 != null)
            msg = msgProc0.apply(msg); 
```

It seems you can access current transaction with the ThreadLocal - `IgniteTxManager#threadCtx`.